### PR TITLE
FixPoint conforms to python `numbers.Real` specification

### PR DESCRIPTION
**Pull request by dpaiton**
addresses #426

### CODE REVIEW

**Comment by dpaiton** [comment]
I have figured out the static check issue, it has to do with the allowable type signatures in `https://github.com/python/typeshed/blob/main/stdlib/unittest/case.pyi#L192`. It's a little confusing still -- I don't know why FixedPoint is not passing the `SupportsRound`, `SupportsAbs`, or `SupportsRSub` checks. I'll file a ticket to resolve that later. Whenever the check does not raise an error, it is because the type is not known (e.g. `expected_result.without_fee_or_slippage` in `test_calc_out_given_in_fp.py` has type `Any`, which is honestly also a bug). The solution is to use `delta` instead of `places`. Then a different type signature is used, which has different requirements. This makes sense anyways, since `places` refers to decimal places, which `unittest` might assume is always passing for FixedPoint. (if it doesn't then it still doesn't make sense).

TL;DR: changing `places` to `delta` will fix the static errors. I'll do this before merging.

File: tests/fixed_point/test_fixed_point.py [hunk 1]
```diff
@@ -322,10 +322,29 @@ def test_modulo_fail(self):
...
+        assert math.floor(FixedPoint("-2.1")) == FixedPoint("-3.0")
+        assert math.floor(FixedPoint("3.6")) == FixedPoint("3.0")
+        assert math.floor(FixedPoint("0.5")) == FixedPoint(0)
+        assert math.floor(FixedPoint(3)) == FixedPoint(0)
+        assert math.floor(FixedPoint(-6)) == FixedPoint("-1.0")
+        assert math.floor(FixedPoint(-6.0)) == FixedPoint(-6.0)
+        assert math.floor(FixedPoint(-6.8)) == FixedPoint(-7.0)
+
+    def test_ceil(self):
+        r"""Test floor operator"""
```
Comment (sentilesdal) [inline tests/fixed_point/test_fixed_point.py]: Test ceil operator

Comment (dpaiton) [reply tests/fixed_point/test_fixed_point.py]: Fixed in a later commit

File: elfpy/math/fixed_point.py [hunk 2]
```diff
@@ -333,14 +333,24 @@ def __ge__(self, other: FixedPoint) -> bool:
...
+
     def __floor__(self) -> FixedPoint:
         """Returns an integer rounded following Python `math.floor` behavior
 
         Given a real number x, return as output the greatest integer less than or equal to x.
         """
         if not self.is_finite():
             return self
-        lhs, rhs = str(self).split(".")
+        lhs, rhs = str(self).split(".")  # lhs = integer part, rhs = fractional part
```
Comment (sentilesdal) [inline elfpy/math/fixed_point.py:334]: could just call the variables themselves `integer, remainder`

File: elfpy/math/fixed_point.py [hunk 3]
```diff
@@ -353,7 +363,7 @@ def __ceil__(self) -> FixedPoint:
...
```
Comment (sentilesdal) [inline elfpy/math/fixed_point.py:347]: could just call the variables themselves integer, remainder

File: elfpy/math/fixed_point.py [hunk 4]
```diff
@@ -372,6 +372,45 @@ def __ceil__(self) -> FixedPoint:
...
 
+    def __round__(self, n: int = 0) -> FixedPoint:
+        """Returns a number rounded following Python `round` behavior.
+
+        Given a real number x and an optional integer n, return as output the number rounded to the closest multiple of 10 to the power -n.
+        If n is omitted, it defaults to 0 (round to nearest integer). Uses Python's "round half to even" strategy.
+        """
+        if not self.is_finite():
+            return self
+        lhs, rhs = str(self).split(".")  # lhs = integer part, rhs = fractional part
```
Comment (sentilesdal) [inline elfpy/math/fixed_point.py:366]: could just call the variables themselves `integer, remainder`

File: elfpy/math/fixed_point.py [hunk 5]
```diff
@@ -372,6 +372,45 @@ def __ceil__(self) -> FixedPoint:
...
+        digit = int(rhs[n])
+        if n == 0 or len(rhs) < n:
+            left_digit = int(lhs[-1])
+        else:
+            left_digit = int(rhs[n - 1])
+        # If these conditions are met, we should round down
+        if digit < 5 or (  # digit less than 5 OR
+            digit == 5  # digit is exactly 5 AND
+            and all(d == "0" for d in rhs[n + 1 :])  # all of the following digits are zero AND
+            and (left_digit % 2 == 0)  # the digit to the left is even
```
Comment (sentilesdal) [inline elfpy/math/fixed_point.py:380]: why does this matter?  it seems crazy to me that you'd round 1.75 to 1.8 and 1.45 to 1.4.  Is that really the standard?

**Review by sentilesdal — APPROVED** [review]
amazing!

File: elfpy/math/fixed_point.py (hunk 5 above)
Comment (dpaiton) [reply elfpy/math/fixed_point.py:380]: Yeah that's how python does rounding (round to even) https://realpython.com/python-rounding/#pythons-built-in-round-function

Comment (dpaiton) [reply elfpy/math/fixed_point.py:380]: we can add args later to implement other methods: https://realpython.com/python-rounding/#a-menagerie-of-methods

File: elfpy/math/fixed_point.py [hunk 6]
```diff
@@ -183,6 +175,18 @@ def __floordiv__(self, other: int | FixedPoint) -> FixedPoint:
             FixedPointIntegerMath.div_down(self.int_value, other.int_value), self.decimal_places, self.signed
         )
 
+    def __rtruediv__(self, other):
+        """Enables reciprocal division to support other / FixedPoint"""
+        return self.__truediv__(other)
+
+    def __floordiv__(self, other: int | FixedPoint) -> FixedPoint:
```
[the remark this answers is not in the document]
Comment (dpaiton) [reply elfpy/math/fixed_point.py:170]: Yeah, for truediv we will call divdown (so whenever we see divdown in solidity we would use `/` in python). And floordiv truncates to "integer", i.e. does divdown (rounds down with 18 decimals) and then we use `floor` to truncate the 18 decimals to just the whole part.

File: elfpy/math/fixed_point.py [hunk 7]
```diff
@@ -241,11 +245,15 @@ def __neg__(self) -> FixedPoint:
         return FixedPoint("-1.0") * self
 
     def __abs__(self) -> FixedPoint:
-        """Enables 'abs()' function"""
+        r"""Enables 'abs()' function"""
```
[the remark this answers is not in the document]
Comment (dpaiton) [reply elfpy/math/fixed_point.py:233]: I might be wrong here, but I think it will make a difference for the backticks to render as an inline code block. If the docstring has only text you don't need `r`, but if it has any sort of sphinx or markdown or whatever then we should use `r`. And since it doesn't really hurt to use it I think it's fine to just use it everywhere.

File: elfpy/math/fixed_point.py [hunk 8]
```diff
@@ -150,7 +153,7 @@ def __rmul__(self, other: int | FixedPoint) -> FixedPoint:
     def __truediv__(self, other):
         """Enables '/' syntax.
 
-        We mirror floordiv because most solidity contract equations use divdown
+        We use div_down to match the majority of Hyperdrive solidity contract equations
```
[the remark this answers is not in the document]
Comment (dpaiton) [reply elfpy/math/fixed_point.py:148]: We want to use the syntax sugar whenever possible. That being said, if a situation comes up where it really makes sense to call it `div_down` then I think it's ok to add it to `FixedPoint`. Notice we already have an example because we have an explicit `div_up`.

File: elfpy/math/fixed_point.py [hunk 9]
```diff
@@ -387,44 +353,44 @@ def __ceil__(self) -> FixedPoint:
                 return FixedPoint(str(int(lhs) + 1) + ".0")  # increase integer component by one
         return FixedPoint(lhs + ".0")  # the number has no remainder
 
-    def __round__(self, n: int = 0) -> FixedPoint:  # pylint: disable=invalid-name
+    def __round__(self, ndigits: int = 0) -> FixedPoint:
```
[the remark this answers is not in the document]
Comment (dpaiton) [reply elfpy/math/fixed_point.py:356]: When I first wrote it I followed python `float` type, which uses `n` in the `__round__` function. However, after looking around more, I found that a lot of other custom types use `ndigits`, which is what is used by the python [round](https://docs.python.org/3/library/functions.html#round) built-in function.
